File: old/tags/2.1/cpp/DIS/TransferControlRequestPdu.cpp
#include <DIS/TransferControlRequestPdu.h> 

using namespace DIS;


TransferControlRequestPdu::TransferControlRequestPdu() : EntityManagementFamilyPdu(),
   _orginatingEntityID(), 
   _recevingEntityID(), 
   _requestID(0), 
   _requiredReliabilityService(0), 
   _tranferType(0), 
   _transferEntityID(), 
   _numberOfRecordSets(0)
{
    setPduType( 35 );
}

TransferControlRequestPdu::~TransferControlRequestPdu()
{
    _recordSets.clear();
}

EntityID& TransferControlRequestPdu::getOrginatingEntityID() 
{
    return _orginatingEntityID;
}

const EntityID& TransferControlRequestPdu::getOrginatingEntityID() const
{
    return _orginatingEntityID;
}

void TransferControlRequestPdu::setOrginatingEntityID(const EntityID &pX)
{
    _orginatingEntityID = pX;
}

EntityID& TransferControlRequestPdu::getRecevingEntityID() 
{
    return _recevingEntityID;
}

const EntityID& TransferControlRequestPdu::getRecevingEntityID() const
{
    return _recevingEntityID;
}

void TransferControlRequestPdu::setRecevingEntityID(const EntityID &pX)
{
    _recevingEntityID = pX;
}

unsigned int TransferControlRequestPdu::getRequestID() const
{
    return _requestID;
}

void TransferControlRequestPdu::setRequestID(unsigned int pX)
{
    _requestID = pX;
}

unsigned char TransferControlRequestPdu::getRequiredReliabilityService() const
{
    return _requiredReliabilityService;
}

void TransferControlRequestPdu::setRequiredReliabilityService(unsigned char pX)
{
    _requiredReliabilityService = pX;
}

unsigned char TransferControlRequestPdu::getTranferType() const
{
    return _tranferType;
}

void TransferControlRequestPdu::setTranferType(unsigned char pX)
{
    _tranferType = pX;
}

EntityID& TransferControlRequestPdu::getTransferEntityID() 
{
    return _transferEntityID;
}

const EntityID& TransferControlRequestPdu::getTransferEntityID() const
{
    return _transferEntityID;
}

void TransferControlRequestPdu::setTransferEntityID(const EntityID &pX)
{
    _transferEntityID = pX;
}

unsigned char TransferControlRequestPdu::getNumberOfRecordSets() const
{
   return _recordSets.size();
}

std::vector<RecordSet>& TransferControlRequestPdu::getRecordSets() 
{
    return _recordSets;
}

const std::vector<RecordSet>& TransferControlRequestPdu::getRecordSets() const
{
    return _recordSets;
}

void TransferControlRequestPdu::setRecordSets(const std::vector<RecordSet>& pX)
{
     _recordSets = pX;
}

void TransferControlRequestPdu::marshal(DataStream& dataStream) const
{
    EntityManagementFamilyPdu::marshal(dataStream); // Marshal information in superclass first
    _orginatingEntityID.marshal(dataStream);
    _recevingEntityID.marshal(dataStream);
    dataStream << _requestID;
    dataStream << _requiredReliabilityService;
    dataStream << _tranferType;
    _transferEntityID.marshal(dataStream);
    dataStream << ( unsigned char )_recordSets.size();

     for(size_t idx = 0; idx < _recordSets.size(); idx++)
     {
        RecordSet x = _recordSets[idx];
        x.marshal(dataStream);
     }

}

void TransferControlRequestPdu::unmarshal(DataStream& dataStream)
{
    EntityManagementFamilyPdu::unmarshal(dataStream); // unmarshal information in superclass first
    _orginatingEntityID.unmarshal(dataStream);
    _recevingEntityID.unmarshal(dataStream);
    dataStream >> _requestID;
    dataStream >> _requiredReliabilityService;
    dataStream >> _tranferType;
    _transferEntityID.unmarshal(dataStream);
    dataStream >> _numberOfRecordSets;

     _recordSets.clear();
     for(size_t idx = 0; idx < _numberOfRecordSets; idx++)
     {
        RecordSet x;
        x.unmarshal(dataStream);
        _recordSets.push_back(x);
     }
}


bool TransferControlRequestPdu::operator ==(const TransferControlRequestPdu& rhs) const
 {
     bool ivarsEqual = true;

     ivarsEqual = EntityManagementFamilyPdu::operator==(rhs);

     if( ! (_orginatingEntityID == rhs._orginatingEntityID) ) ivarsEqual = false;
     if( ! (_recevingEntityID == rhs._recevingEntityID) ) ivarsEqual = false;
     if( ! (_requestID == rhs._requestID) ) ivarsEqual = false;
     if( ! (_requiredReliabilityService == rhs._requiredReliabilityService) ) ivarsEqual = false;
     if( ! (_tranferType == rhs._tranferType) ) ivarsEqual = false;
     if( ! (_transferEntityID == rhs._transferEntityID) ) ivarsEqual = false;

     for(size_t idx = 0; idx < _recordSets.size(); idx++)
     {
        if( ! ( _recordSets[idx] == rhs._recordSets[idx]) ) ivarsEqual = false;
     }


    return ivarsEqual;
 }

int TransferControlRequestPdu::getMarshalledSize() const
{
   int marshalSize = 0;

   marshalSize = EntityManagementFamilyPdu::getMarshalledSize();
   marshalSize = marshalSize + _orginatingEntityID.getMarshalledSize();  // _orginatingEntityID
   marshalSize = marshalSize + _recevingEntityID.getMarshalledSize();  // _recevingEntityID
   marshalSize = marshalSize + 4;  // _requestID
   marshalSize = marshalSize + 1;  // _requiredReliabilityService
   marshalSize = marshalSize + 1;  // _tranferType
   marshalSize = marshalSize + _transferEntityID.getMarshalledSize();  // _transferEntityID
   marshalSize = marshalSize + 1;  // _numberOfRecordSets

   for(int idx=0; idx < _recordSets.size(); idx++)
   {
        RecordSet listElement = _recordSets[idx];
        marshalSize = marshalSize + listElement.getMarshalledSize();
    }

    return marshalSize;
}

File: old/tags/2.1/java/edu/nps/moves/dis/TransferControlRequestPdu.java
package edu.nps.moves.dis;

import java.util.*;
import java.io.*;
import edu.nps.moves.jaxb.dis.*;

/**
 * Section 5.3.9.3 Information initiating the dyanic allocation and control of simulation entities         between two simulation applications. Requires manual cleanup. The padding between record sets is variable. UNFINISHED
 *
 * Copyright (c) 2007, MOVES Institute, Naval Postgraduate School. All rights reserved.
 * This work is licensed under the BSD open source license, available at https://www.movesinstitute.org/licenses/bsd.html
 *
 * @author DMcG
 */
public class TransferControlRequestPdu extends EntityManagementFamilyPdu implements Serializable
{
   /** ID of entity originating request */
   protected EntityID  orginatingEntityID = new EntityID(); 

   /** ID of entity receiving request */
   protected EntityID  recevingEntityID = new EntityID(); 

   /** ID ofrequest */
   protected long  requestID;

   /** required level of reliabliity service. */
   protected short  requiredReliabilityService;

   /** type of transfer desired */
   protected short  tranferType;

   /** The entity for which control is being requested to transfer */
   protected EntityID  transferEntityID = new EntityID(); 

   /** number of record sets to transfer */
   protected short  numberOfRecordSets;

   /** @@@This is wrong--the RecordSet class needs more work */
   protected List recordSets = new ArrayList(); 

/** Constructor */
 public TransferControlRequestPdu()
 {
    setPduType( (short)35 );
 }

/** 
 * Constructor--takes a parallel jaxb object and returns an open-dis object 
 * 1.4_sed_bait_start */
 public TransferControlRequestPdu(edu.nps.moves.jaxb.dis.TransferControlRequestPdu x)
 {
     super(x); // Call superclass constructor


     edu.nps.moves.dis.EntityID foo_0;
     if(x.getOrginatingEntityID() == null)
        foo_0 = new edu.nps.moves.dis.EntityID();
      else
        foo_0 = new edu.nps.moves.dis.EntityID(x.getOrginatingEntityID() );
     this.setOrginatingEntityID(foo_0);


     edu.nps.moves.dis.EntityID foo_1;
     if(x.getRecevingEntityID() == null)
        foo_1 = new edu.nps.moves.dis.EntityID();
      else
        foo_1 = new edu.nps.moves.dis.EntityID(x.getRecevingEntityID() );
     this.setRecevingEntityID(foo_1);

     this.requestID = x.getRequestID();
     this.requiredReliabilityService = x.getRequiredReliabilityService();
     this.tranferType = x.getTranferType();

     edu.nps.moves.dis.EntityID foo_5;
     if(x.getTransferEntityID() == null)
        foo_5 = new edu.nps.moves.dis.EntityID();
      else
        foo_5 = new edu.nps.moves.dis.EntityID(x.getTransferEntityID() );
     this.setTransferEntityID(foo_5);

     this.numberOfRecordSets = x.getNumberOfRecordSets();
     this.recordSets = new ArrayList();
     for(int idx = 0; idx < x.getRecordSets().size(); idx++)
     {
        this.recordSets.add( new edu.nps.moves.dis.RecordSet((edu.nps.moves.jaxb.dis.RecordSet) x.getRecordSets().get(idx)));
     }
 }
/* 1.4_sed_bait_end */


/**
 * returns a jaxb object intialized from this object, given an empty jaxb object
 * 1.4_sed_bait_start **/
 public edu.nps.moves.jaxb.dis.TransferControlRequestPdu initializeJaxbObject(edu.nps.moves.jaxb.dis.TransferControlRequestPdu x)
 {
     super.initializeJaxbObject(x); // Call superclass initializer

     ObjectFactory factory = new ObjectFactory();

     x.setOrginatingEntityID( this.getOrginatingEntityID().initializeJaxbObject(factory.createEntityID()) );
     x.setRecevingEntityID( this.getRecevingEntityID().initializeJaxbObject(factory.createEntityID()) );
     x.setRequestID( this.getRequestID() );
     x.setRequiredReliabilityService( this.getRequiredReliabilityService() );
     x.setTranferType( this.getTranferType() );
     x.setTransferEntityID( this.getTransferEntityID().initializeJaxbObject(factory.createEntityID()) );
     x.setNumberOfRecordSets( this.getNumberOfRecordSets() );

     List recordSets_1 = x.getRecordSets();
     for(int idx = 0; idx < recordSets.size(); idx++)
     {
         RecordSet a = (edu.nps.moves.dis.RecordSet)recordSets.get(idx);
         recordSets_1.add(a.initializeJaxbObject(factory.createRecordSet()));
     }
   return x;
 }
/* 1.4_sed_bait_end */


public int getMarshalledSize()
{
   int marshalSize = 0; 

   marshalSize = super.getMarshalledSize();
   marshalSize = marshalSize + orginatingEntityID.getMarshalledSize();  // orginatingEntityID
   marshalSize = marshalSize + recevingEntityID.getMarshalledSize();  // recevingEntityID
   marshalSize = marshalSize + 4;  // requestID
   marshalSize = marshalSize + 1;  // requiredReliabilityService
   marshalSize = marshalSize + 1;  // tranferType
   marshalSize = marshalSize + transferEntityID.getMarshalledSize();  // transferEntityID
   marshalSize = marshalSize + 1;  // numberOfRecordSets
   for(int idx=0; idx < recordSets.size(); idx++)
   {
        RecordSet listElement = (RecordSet)recordSets.get(idx);
        marshalSize = marshalSize + listElement.getMarshalledSize();
   }

   return marshalSize;
}


public void setOrginatingEntityID(EntityID pOrginatingEntityID)
{ orginatingEntityID = pOrginatingEntityID;
}

public EntityID getOrginatingEntityID()
{ return orginatingEntityID; }

public void setRecevingEntityID(EntityID pRecevingEntityID)
{ recevingEntityID = pRecevingEntityID;
}

public EntityID getRecevingEntityID()
{ return recevingEntityID; }

public void setRequestID(long pRequestID)
{ requestID = pRequestID;
}

public long getRequestID()
{ return requestID; 
}

public void setRequiredReliabilityService(short pRequiredReliabilityService)
{ requiredReliabilityService = pRequiredReliabilityService;
}

public short getRequiredReliabilityService()
{ return requiredReliabilityService; 
}

public void setTranferType(short pTranferType)
{ tranferType = pTranferType;
}

public short getTranferType()
{ return tranferType; 
}

public void setTransferEntityID(EntityID pTransferEntityID)
{ transferEntityID = pTransferEntityID;
}

public EntityID getTransferEntityID()
{ return transferEntityID; }

public short getNumberOfRecordSets()
{ return (short)recordSets.size();
}

/** Note that setting this value will ot change the marshalled value. The list whose length this describes is used for that purpose.
 * The getnumberOfRecordSets method will also be based on the actual list length rather than this value. 
 * The method is simply here for java bean completeness.
 */
public void setNumberOfRecordSets(short pNumberOfRecordSets)
{ numberOfRecordSets = pNumberOfRecordSets;
}

public void setRecordSets(List pRecordSets)
{ recordSets = pRecordSets;
}

public List getRecordSets()
{ return recordSets; }


public void marshal(DataOutputStream dos)
{
    super.marshal(dos);
    try 
    {
       orginatingEntityID.marshal(dos);
       recevingEntityID.marshal(dos);
       dos.writeInt( (int)requestID);
       dos.writeByte( (byte)requiredReliabilityService);
       dos.writeByte( (byte)tranferType);
       transferEntityID.marshal(dos);
       dos.writeByte( (byte)recordSets.size());

       for(int idx = 0; idx < recordSets.size(); idx++)
       {
            RecordSet aRecordSet = (RecordSet)recordSets.get(idx);
            aRecordSet.marshal(dos);
       } // end of list marshalling

    } // end try 
    catch(Exception e)
    { 
      System.out.println(e);}
    } // end of marshal method

public void unmarshal(DataInputStream dis)
{
    super.unmarshal(dis);

    try 
    {
       orginatingEntityID.unmarshal(dis);
       recevingEntityID.unmarshal(dis);
       requestID = dis.readInt();
       requiredReliabilityService = dis.readByte();
       tranferType = dis.readByte();
       transferEntityID.unmarshal(dis);
       numberOfRecordSets = dis.readByte();
        for(int idx = 0; idx < numberOfRecordSets; idx++)
        {
           RecordSet anX = new RecordSet();
            anX.unmarshal(dis);
            recordSets.add(anX);
        };

    } // end try 
   catch(Exception e)
    { 
      System.out.println(e); 
    }
 } // end of unmarshal method 


 /**
  * The equals method doesn't always work--mostly on on classes that consist only of primitives. Be careful.
  */
 public boolean equals(TransferControlRequestPdu rhs)
 {
     boolean ivarsEqual = true;

    if(rhs.getClass() != this.getClass())
        return false;

     if( ! (orginatingEntityID.equals( rhs.orginatingEntityID) )) ivarsEqual = false;
     if( ! (recevingEntityID.equals( rhs.recevingEntityID) )) ivarsEqual = false;
     if( ! (requestID == rhs.requestID)) ivarsEqual = false;
     if( ! (requiredReliabilityService == rhs.requiredReliabilityService)) ivarsEqual = false;
     if( ! (tranferType == rhs.tranferType)) ivarsEqual = false;
     if( ! (transferEntityID.equals( rhs.transferEntityID) )) ivarsEqual = false;
     if( ! (numberOfRecordSets == rhs.numberOfRecordSets)) ivarsEqual = false;

     for(int idx = 0; idx < recordSets.size(); idx++)
     {
        RecordSet x = (RecordSet)recordSets.get(idx);
        if( ! ( recordSets.get(idx).equals(rhs.recordSets.get(idx)))) ivarsEqual = false;
     }


    return ivarsEqual;
 }
} // end of class
